cpp/tests/rolling/rolling_test.cu

```diff
@@ -0,0 +1,502 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <tests/utilities/cudf_test_fixtures.h>
+#include <tests/utilities/cudf_test_utils.cuh>
+
+#include <rolling.hpp>
+#include <cudf.h>
+
+#include <utilities/error_utils.hpp>
+#include <utilities/cudf_utils.h>
+#include "tests/utilities/column_wrapper.cuh"
+
+#include <groupby/aggregation_operations.hpp>
+
+#include <gtest/gtest.h>
+
+#include <vector>
+#include <random>
+#include <algorithm>
+#include <memory>
+
+template <typename T>
+class RollingTest : public GdfTest {
+
+protected:
+  // only for integral types
+  template <typename U = T, typename std::enable_if_t<std::is_integral<U>::value, std::nullptr_t> = nullptr>
+  void make_random_input(size_t nrows, bool add_nulls)
```

Comment:
I think your testing code could be greatly simplified and remove a lot of duplicate functionality by using `cudf::test::column_wrapper`. For example see: https://github.com/rapidsai/cudf/blob/c81152e424871b5f00a777d6bca6ccde3301f536/cpp/tests/copying/scatter_tests.cu#L215-L217

You can easily pass a lambda generator for data and nulls to the column_wrapper, eliminating this whole `make_random_input` function and a lot of other stuff you've written here.

Reply:
I replaced `gdf_col_pointer` with `column_wrapper` for input data. I still have `make_random_input` since it's a convenient function to generate random data on the host. I'm not sure how `column_wrapper` would replace it since it still requires some data generator. I could move the lambdas from `make_random_input` to the typed test body, but then there is going to be a lot of repeating code in the test definitions - it's more readable to have a single call to `make_random_input` in the test body.

Reply:
It's just that every single module duplicates this functionality. The duplication across our tests gives me heartburn. :)